_posts/2020-01-31-Everything-you-need-to-know-about-ISP.md

```diff
@@ -0,0 +1,166 @@
+A bulky interface might be a good indicator for ISP violation, although it's not always the case
+
+2. UNSupportedMethod Exception
+As in our burger example, if you encounter too many **Unsupported Operation Exception** it could be a design problem related to ISP. It might be a good time to refactor these classes, for example, we can refactor our BurgerPoint code to have separate interfaces for Burger order and Fries order.
+```java
+/**
+ * IBurgerOrder.java
+ **/
+public interface IBurgerOrder{
+    void orderBurger(int quantity);
+}
+```
+```java
+/**
+ * IFriesOrder.java
+ **/
+public interface IFriesOrder {
+    void orderFries(int quantity);
+}
+```
+You could also use [Adapter pattern](https://en.wikipedia.org/wiki/Adapter_pattern). Adapter patten is used to abstract the unwanted methods like below:
+```java
+/**
+ * IAdapterOrderForBurger.java
+ **/
+interface IAdapterOrderForBurger {
+    void orderBurger(int quantity);
+}
+```
+```java
+/**
+ * IAdapterOrderForFries.java
+ **/
+interface IAdapterOrderForFries {
+    void orderFries(int quantity);
+}
+```
+We have created independent interfaces which will only have the methods needed from the **IOrder.java** interface
+```java
+/**
+ * AdaptedBurgerOrder.java
+ **/
+public class AdaptedBurgerOrder implements IAdapterOrderForBurger {
+    private IOrder burgerOrder;
+    public AdaptedBurgerOrder(IOrder burgerOrder){
+        this.burgerOrder = burgerOrder;
+    }
+
+    @Override
+    public void orderBurger(int quantity) {
+        burgerOrder.orderBurger(quantity);
+    }
+}
+// Similarly for fries order
+```
+As we can see in above code, we are still using the methods provided by **IOrder.java** interface but we are only implementing the methods that the concrete class actually needs.
+This is useful when you are using an external dependency. In the above example  we are using order interface as an external dependency but we have successfully restructured code to avoid side effects. 
+
+3. Too many dependencies to mock for JUnit    
+If your class have too many dependencies while methods only needed few, this might indicate for ISP violation which leads to minimal dependency
+
+4. Too many wrapper classes    
+As wrapper classes add no additional functionality in the code rather is a design choice, having too many of them indicates for refactoring of the codebase. Having too many of wrapper classes will also need maintenance and further increases complexity in codebase
+
+## Why always only having a single method interface is not a good idea?
+Applying ISP to the extreme will result in one method interfaces also known as **role-interfaces**. 
+
+This solution will definitely solve the problem for ISP violation but it can result in SRP (Single Responsibility Principle) violation. For example, the **Collection** interface in java has more than one methods e.g. **size()** and **isEmpty()** which are support methods and it make sense for them to be in a single interface
```

Comment:
Although I follow a different interpretation of SRP, I believe it's a valid point since a class can have multiple role interfaces implemented and ISP is more about the client-side and SRP is more about implementation side. Thanks for pointing out 

![image](https://user-images.githubusercontent.com/9193291/75249908-3245d080-57fd-11ea-8a2c-da506ed6f5c0.png)

https://blog.cleancoder.com/uncle-bob/2014/05/08/SingleReponsibilityPrinciple.html

I would also say if the SRP would be a princile that is also defined on interfaces then role interfaces might violdate SRP. I would say that ISP is aggressive on separating and SRP is more about balance.

and interestingly ISP might even be redundant in that case.